File: neorg_lsp/rpc/completion.py
import attrs

from neorg_lsp.rpc import BaseRequest, BaseResponse
from neorg_lsp.state import State


@attrs.define
class CompletionPosition:
    line: int
    character: int


@attrs.define
class TextDocumentIdentifier:
    uri: str


@attrs.define
class TextDocumentPositionParams:
    textDocument: TextDocumentIdentifier
    position: CompletionPosition


@attrs.define
class CompletionParams(TextDocumentPositionParams):
    pass


@attrs.define
class CompletionRequest(BaseRequest):
    params: CompletionParams | None = None


@attrs.define
class MarkupContent:
    kind: str  # "plaintext" | "markdown"
    value: str | list[str]


@attrs.define
class CompletionItem:
    label: str
    documentation: str | MarkupContent | None = None


@attrs.define
class CompletionItemResult(BaseResponse):
    result: list[CompletionItem] | None = None

    @classmethod
    def get_completions(cls, req: CompletionRequest, lspState: State):
        def all_docs(doc_uris: list[str]) -> str:
            return "\n\t".join(
                [
                    f"[{lspState.documents[doc_uri].title}]({doc_uri})"
                    for doc_uri in doc_uris
                ]
            )

        return cls(
            req.id,
            result=[
                CompletionItem(
                    category,
                    MarkupContent(
                        "markdown",
                        f"""
# {category}

Found in the following files:

[
    {all_docs(category_docs.uris)}
]
""",
                    ),
                )
                for category, category_docs in lspState.labels.items()
            ],
        )

File: CeorgLsp/Rpc/Completion.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CeorgLsp.Rpc
{
    public readonly record struct CompletionRequestParams
    {
        [JsonPropertyName("textDocument")]
        public required Uri TextDocument { get; init; }

        [JsonPropertyName("postion")]
        public required Position Postion { get; init; }

        [JsonPropertyName("completionContext")]
        public JsonElement? CompletionContext { get; init; }
    }

    public readonly record struct Position
    {
        [JsonPropertyName("line")]
        public required uint Line { get; init; }

        [JsonPropertyName("character")]
        public required uint Character { get; init; }
    }

    public readonly record struct CompletionResult
    {
        [JsonPropertyName("result")]
        public required CompletionItem[] Result { get; init; }
    }

    public readonly record struct CompletionItem
    {
        [JsonPropertyName("label")]
        public required string Label { get; init; }

        [JsonPropertyName("documentation")]
        public MarkupContent? LabelDetails { get; init; }
    }

    public readonly record struct MarkupContent
    {
        [JsonPropertyName("kind")]
        public required string MarkupKind { get; init; }

        [JsonPropertyName("value")]
        public required string Value { get; init; }
    }

    public enum MarkupKind
    {
        plaintext,
        markdown
    }
}
